plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -2909,147 +2863,143 @@ void producer_plugin_impl::start_write_window() {
       }));
 }
 
-// Called from app thread
+// Called only from app thread
 void producer_plugin_impl::switch_to_read_window() {
    EOS_ASSERT(app().executor().is_write_window(),  producer_exception, "expected to be in write window");
-   EOS_ASSERT(_ro_num_active_trx_exec_tasks.load() == 0 && _ro_trx_exec_tasks_fut.empty(),  producer_exception, "_ro_trx_exec_tasks_fut expected to be empty" );
+   EOS_ASSERT( _ro_num_active_exec_tasks.load() == 0 && _ro_exec_tasks_fut.empty(), producer_exception, "_ro_exec_tasks_fut expected to be empty" );
 
    _ro_timer.cancel();
    _time_tracker.add_idle_time( fc::time_point::now() - _idle_trx_time );
 
    // we are in write window, so no read-only trx threads are processing transactions.
-   // _ro_trx_queue is not being accessed. No need to lock.
-   if ( _ro_trx_queue.empty() ) { // no read-only trxs to process. stay in write window
+   if ( app().executor().read_only_queue().empty() ) { // no read-only tasks to process. stay in write window
       start_write_window(); // restart write window timer for next round
       return;
    }
 
    auto& chain = chain_plug->chain();
    uint32_t pending_block_num = chain.head_block_num() + 1;
-   app().executor().set_to_read_window();
+   _ro_window_deadline = fc::time_point::now() + _ro_read_window_effective_time_us;
+   app().executor().set_to_read_window(_ro_thread_pool_size,
+      [received_block=&_received_block, pending_block_num, ro_window_deadline=_ro_window_deadline]() {
+         return fc::time_point::now() >= ro_window_deadline || (received_block->load() >= pending_block_num); // should_exit()
+      });
    chain.set_db_read_only_mode();
+   _ro_in_read_only_mode = true;
    _ro_read_window_start_time = fc::time_point::now();
    _ro_all_threads_exec_time_us = 0;
 
    // start a read-only transaction execution task in each thread in the thread pool
-   _ro_num_active_trx_exec_tasks = _ro_thread_pool_size;
-   auto start_time = fc::time_point::now();
-   _ro_trx_queue.set_exit_criteria(_ro_thread_pool_size, &_received_block, pending_block_num, start_time + _ro_read_window_effective_time_us);
+   _ro_num_active_exec_tasks = _ro_thread_pool_size;
    for (auto i = 0; i < _ro_thread_pool_size; ++i ) {
```

Comment:
Maybe add `_ro_exec_tasks_fut.resize(0)` before this line, and we won't need the two `self->_ro_exec_tasks_fut.clear();` in the lambda of the async_wait.

Reply:
I prefer the current `clear()` so the futures are destroyed earlier rather than later.

Reply:
I don't really see a benefit to have the futures destroyed a little bit earlier, and it requires to have to check the logic to convince yourself that the `_ro_exec_tasks_fut` vector is empty before we start emplacing futures into it.
Also the resize(0) would have the additional benefit that the vector capacity is preserved, so that we don't reallocate memory every time. 

If you prefer to keep it as is, would you consider adding:

```
   assert(_ro_exec_tasks_fut.empty());
   _ro_exec_tasks_fut.reserve(_ro_thread_pool_size);
```

Reply:
Oh, my bad, I see that `clear()` leaves the vector capacity unchanged. For some reason I thought otherwise. So we're good as is (though I think the `assert()` would be nice :-)!

Reply:
`clear()` is not required to change capacity and, in my experience, most (maybe all) do not change capacity. `std::vector::resize(0)` c++ spec does not say anything about capacity. Although cppreference does say it maintains capacity, it makes that determination because iterators are maintained. But maintaining iterators does not matter if resizing to 0. So I think a compliant implementation could reduce capacity for `resize(0)`. It would have to special case for the value of `0`, but there might be benefit. Regardless, in practice I think `clear()` and `resize(0)` do the same thing for most/all implementations.

We could add the `_ro_exec_tasks_fut.resize(0)` while maintaining the existing `_ro_exec_tasks_fut.clear()` calls. Ok with that?

Reply:
> We could add the `_ro_exec_tasks_fut.resize(0)` while maintaining the existing `_ro_exec_tasks_fut.clear()` calls. Ok with that?

Sure. or the `assert(_ro_exec_tasks_fut.empty());`. I'm fine with either.

Reply:
Done.